rust/integration-testing/src/bin/flight-test-integration-client.rs

```diff
@@ -0,0 +1,439 @@
+        .arg(Arg::with_name("port").long("port").takes_value(true))
+        .arg(Arg::with_name("path").long("path").takes_value(true))
+        .arg(
+            Arg::with_name("scenario")
+                .long("scenario")
+                .takes_value(true),
+        )
+        .get_matches();
+
+    let host = matches.value_of("host").expect("Host is required");
+    let port = matches.value_of("port").expect("Port is required");
+
+    match matches.value_of("scenario") {
+        Some("middleware") => middleware_scenario(host, port).await?,
+        Some("auth:basic_proto") => auth_basic_proto_scenario(host, port).await?,
+        Some(scenario_name) => unimplemented!("Scenario not found: {}", scenario_name),
+        None => {
+            let path = matches
+                .value_of("path")
+                .expect("Path is required if scenario is not specified");
+            integration_test_scenario(host, port, path).await?;
+        }
+    }
+
+    Ok(())
+}
+
+async fn middleware_scenario(host: &str, port: &str) -> Result {
+    let url = format!("http://{}:{}", host, port);
+    let conn = tonic::transport::Endpoint::new(url)?.connect().await?;
+    let mut client = FlightServiceClient::with_interceptor(conn, middleware_interceptor);
+
+    let mut descriptor = FlightDescriptor::default();
+    descriptor.set_type(DescriptorType::Cmd);
+    descriptor.cmd = b"".to_vec();
+
+    // This call is expected to fail.
+    let resp = client.get_flight_info(Request::new(descriptor.clone()));
+
+    let resp = resp.await;
+
+    match resp {
+        Ok(_) => return Err(Box::new(Status::internal("Expected call to fail"))),
+        Err(e) => {
+            let headers = e.metadata();
+            let middleware_header = headers.get("x-middleware");
+            let value = middleware_header.map(|v| v.to_str().unwrap()).unwrap_or("");
+
+            if value != "expected value" {
+                let msg = format!(
+                    "On failing call: Expected to receive header 'x-middleware: expected value', \
+                     but instead got: '{}'",
+                    value
+                );
+                return Err(Box::new(Status::internal(msg)));
+            }
+        }
+    }
+
+    // This call should succeed
+    descriptor.cmd = b"success".to_vec();
+    let resp = client.get_flight_info(Request::new(descriptor)).await?;
+
+    let headers = resp.metadata();
+    let middleware_header = headers.get("x-middleware");
+    let value = middleware_header.map(|v| v.to_str().unwrap()).unwrap_or("");
+
+    if value != "expected value" {
+        let msg = format!(
+            "On success call: Expected to receive header 'x-middleware: expected value', \
+            but instead got: '{}'",
+            value
+        );
+        return Err(Box::new(Status::internal(msg)));
+    }
+
+    Ok(())
+}
+
+fn middleware_interceptor(mut req: Request<()>) -> Result<Request<()>, Status> {
```

Comment:
Was this junk leftover from my experiments, or is this needed?

Reply:
This is needed https://github.com/integer32llc/arrow/pull/4/files#diff-8c6750d1fd1e0371e4b71b357cde69fa2bd622825253108f472e21dc592793a2R79